### PR TITLE
Add an `alias_kind` directive to configure wrapper macros

### LINKED ISSUE #18: Allow wrappers
Migrated from bazelbuild/rules_go#748

Allow people to mark rules as a wrapper for a go rule, so they can be managed by gazelle as if the were the underlying rule.
It would be even nicer if you could do this once for an entire rule type rather than per rule. The original example was bazelbuild/rules_docker#119 where go_image is a proxy for go_binary as far as gazelle should be concerned.

### PR DESCRIPTION
**Pull request by alex-torok**
**What type of PR is this?**

Feature

**What package or component does this PR mostly affect?**

cmd/gazelle

**What does this PR do? Why is it needed?**

Introduces an `alias_kind` directive to instruct gazelle for how to handle wrapper macros. Wrapper macros are starlark macros around rules that gazelle knows how to understand. A wrapper macro will have the same attrs that the underlying rule kind has, but likely does additional things like setting up common deps or generating additional targets.

A common pattern for creating wrapper macros that gazelle should understand is the custom `rule.verb` usecase described here - https://bazel.build/rules/verbs-tutorial.

Gazelle should still be able to understand how to index and update a wrapper macro as long as it is told what underlying rule the wrapper macro should be considered as.

`alias_kind` is unique from `map_kind` in that it does not force a rule into the kind that is specified in the `alias_kind` directive. In the above "verbs" example, someone may configure a wrapper macro around a `py_binary` target, while still wanting to keep other py_binary targets around in their BUILD file. `map_kind` would not allow this, as it would force ALL py_binary targets into the mapped kind.

**Which issues(s) does this PR fix?**

Fixes #18

**Other notes for review**

Getting this functionality required a few changes that I'll discuss here since the map_kind and wrapper macro functionality is similar, yet subtly different:

1. We now pass the wrapper macro config into the rule merger so that it can match an existing wrapper macro against the newly generated rule, even though their kinds are not strictly the same. The mapped_kind functionality did not need to do this, as the kind mapping logic in fix-update.go handles updating the generated rule to match the new kind prior to build file merging. We cannot do something similar here, as the wrapper macro pattern should not force all rules to exist in the new type. Instead, it should instruct the merger that the newly generated rule can match the wrapper macro.
2. Update the metaresolver to perform a lookup in the wrapper macro configuration. When we come across a wrapper macro, we need to use the inverseKindMapper so that we correctly update the existing rule's kind prior to sending it off to a language extension to be indexed.

### CODE REVIEW

File: cmd/gazelle/metaresolver.go
Comment (alex-torok) [inline cmd/gazelle/metaresolver.go:65]: I have a follow-up that refactors the metaresolver to instead take a `c *config.Config` in the `Resolver` method, which removes this and the above `MappedKind` handling. I decided to keep them separate and follow existing patterns in this PR.

File: README.rst
Comment (alex-torok) [inline README.rst:901]: I realize this naming of `# gazelle:macro <NAME> wraps <RULE>` is somewhat cutesy, but I was struggling to decide on the ordering of the wrapper macro name and underlying rule. Being chatty and reading like plain english seemed like a reasonable solution, but I'm open to suggestions.

**Comment by alex-torok** [comment]
@jayconrod - would you be able to review this? It solves a long-outstanding issue with supporting arbitrary wrapper macros around rules that gazelle extensions understand how to index and resolve.

**Comment by jayconrod** [comment]
Sorry for the slow response -- I'm finishing up some projects due by the end of the quarter, but I hope to be able to review and give feedback in the next day or two.

I'm not an active maintainer at the moment, so the folks who are will have the final say.

**Comment by jayconrod** [comment]
Sorry to take so long to review. The end of the year continues to be busy. I'm hoping to have more time for open source work next quarter.

Before getting too deep into the implementation, I want to understand the behavior and user experience. This seems similar enough to `# gazelle:map_kind` that it's likely to be confusing to users (it's confusing to me anyway). As I understand:

- `# gazelle:map_kind go_binary go_binary_macro //:def.bzl`
  - Instead of generating a new `go_binary` target, Gazelle generates a `go_binary_macro` call, loading from `//:def.bzl`.
  - Gazelle also replaces existing `go_binary` targets with `go_binary_macro`.
- `# gazelle:macro go_binary_macro wraps go_binary`
  - This directive does not affect new target generation, so Gazelle still generates `go_binary`, not `go_binary_macro`.
  - When there is an existing `go_binary_macro` target, Gazelle updates (or indexes) that instead of `go_binary`.

Assuming that's correct (please correct me if not), here are my thoughts:

- First, I'm curious about your use case for wrapper macros. I was skeptical of the initial `map_kind` feature. It feels like it shouldn't be needed if Gazelle is smart enough to infer what you want and if your project follows conventions closely enough to make that possible. But the Bazel world is messy, so I'm curious to hear examples where that doesn't work.
- Since this sounds so similar to `map_kind`, I'd suggest giving the directive a similar name. How about `alias_kind`?

    ```
    # gazelle:alias_kind go_binary_macro go_binary
    ```

- I think it's better without the `wraps` by the way. Words in directives are like positional arguments in a command, and it feels strange to me to have `wraps` in there just for readability. It's very different than other directives.

**Comment by alex-torok** [comment]
Your understanding of the desired functionality is correct. This new macro would not change how rules are generated, but allow users to manually change the kind of a rule, while still letting gazelle update the deps.

> I'm curious about your use case for wrapper macros.

A pattern that we have a lot in our repo is people setting up wrapper macros to abstract wiring up common target patterns. As an example, you may have a wrapper macro for defining a python binary that gets packaged into docker image with some additional binaries or data. We may have many of these image targets and want to share some invariant configuration across all of them. A macro would be written to handle the boilerplate of wiring multiple targets.

Another more concrete usecase that is seen in the wild is the example of having a macro create custom `<name>.<verb>` targets - https://bazel.build/rules/verbs-tutorial.

In both of these cases, the wrapper macro creates the underlying rule that gazelle knows how to generate/index/update and uses the same attr values that are passed to that underlying rule.

So far, I've solved this issue by implementing wrappers around each language extension to handle, but it is cumbersome and difficult for non-gazelle-experts in my company to see what is going on and extend new wrapper macros.

Also, this issue has come up a few times on bazel gazelle slack -- [one](https://bazelbuild.slack.com/archives/C01HMGN77Q8/p1657162495641359), [two](https://bazelbuild.slack.com/archives/C01HMGN77Q8/p1724354252411379) (admittedly this second one is me, but others chimed in)

I could definitely expand the docs to clarify the differences between map_kind and wrapper macro handling.

> Since this sounds so similar to map_kind, I'd suggest giving the directive a similar name. How about alias_kind?

I like alias_kind - I'll update the PR to use that and drop the `wraps` word. I agree it doesn't really follow the existing patterns and `alias_kind` makes the ordering clear since it would follow `map_kind`'s

File: cmd/gazelle/update-repos.go
Comment (jayconrod) [inline cmd/gazelle/update-repos.go:310]: NIt: can you pass in `nil` instead of allocating a `map[string]string{}`? Read operations on a `nil` map work the same as an empty map, it's just the write operations that panic, and I don't think `MergeFile` needs to write this map.

File: merger/merger.go
Comment (alex-torok) [reply merger/merger.go:109]: Whoops... I had a feeling that I missed something. Thanks for pointing this out.

File: config/config.go
Comment (alex-torok) [reply config/config.go:303]: In my mind, I don't see a big gain for adding the load file path. Someone would have to update a rule to be the alias kind, then re-run gazelle just to have it update the loads. 

I do agree that it would be rare for same-named aliases to exist in multiple different files. 

Either way, supporting that functionality seems like something that could eventually be done in a follow-up if someone really wants it. I could see this functionality working with alias_kind taking two or three arguments.

File: merger/merger_test.go
Comment (alex-torok) [reply merger/merger_test.go:948]: Sorry for sending out a bit of a sloppy review request. I'll address this at some point this week and add the test I was planning on adding.

File: cmd/gazelle/integration_test.go
Comment (alex-torok) [reply cmd/gazelle/integration_test.go:2732]: Do you mean a case where the wrapper macro has the same name as the underlying rule?

```
# gazelle:alias_kind foo foo
```

In my mind that should be an error (I can update the config parsing logic to warn when that happens).

File: config/config.go
Comment (jayconrod) [reply config/config.go:303]: Seems reasonable, agree that it could be extended if needed.

File: cmd/gazelle/integration_test.go
Comment (jayconrod) [reply cmd/gazelle/integration_test.go:2732]: If it's an error, then it's reasonable not to test it here. We can relax that later (and then test) if needed.

**Review by jayconrod — APPROVED** [review]
All looks good. Thanks for doing this!

**Comment by jayconrod** [comment]
Tagging @linzhp @fmeum as reviewers / approvers.

**Pull request merged**
